### PR TITLE
Resteasy Reactive: Support use of optional with list/set/sortedset

### LINKED ISSUE #23898: Can't use Optional with List with @QueryParam with reasteasy controller
### Describe the bug

When using @QueryParam("name") Optional<List<Long>> in a resteasy reactive controller, we get an error:

```
//
Caused by: java.lang.RuntimeException: Failed to find converter for java.util.List 
//
```

When used without Optional we don't get this error. 

This works: 

```java

    @GET
    @Path("/table")
    @Transactional
    public Page<Stuff> getStuff(
        @QueryParam("test") final List<Long> ids,
        @BeanParam PageReq page
    ) {
       //
    }
```

### Expected behavior

We can use an Optional<List<Long>> in a resteasy reactive controller

### Actual behavior

We get a compilation error

### How to Reproduce?

1. create a Quarkus project, with reasteasy-reactive extension
2. create a controller with the following endpoint

```java

    @GET
    @Path("/table")
    @Transactional
    public Page<Stuff> getStuff(
        @QueryParam("test") final Optional<List<Long>> ids,
        @BeanParam PageReq page
    ) {
       //
    }
```

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

tested with 2.6.3.Final and tested in 2.7.2.Final 

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by Sgitario**
Support use when using query params with some collections. Example:

```java
@Path("/list")
    @GET
    public String sayHelloToList(@QueryParam("name") final Optional<List<String>> names) {
        return doSayHelloToCollection(names);
    }
```

fix https://github.com/quarkusio/quarkus/issues/23898

### CODE REVIEW

**Comment by Sgitario** [comment]
Moving to Draft as this is not fully working yet.

**Review by geoand — APPROVED** [review]
Thanks!

**Comment by Sgitario** [comment]
> Moving to Draft as this is not fully working yet.

Now, the PR is ready to be reviewed.

This is a controversial difference of behaviors here between Resteasy Reactive and Resteasy Classic. Let's imagine we have the following endpoint:

```java
@Path("/list")
@GET
public String sayHelloToList(@QueryParam("name") final Optional<List<String>> names) {
   // ...
}
```

If no query params are being set, I would expect to receive an optional empty. This is working like this in Resteasy Reactive, but in Resteasy Classic, you'll receive the optional with an empty list of strings.

**Comment by geoand** [comment]
`Optional<List<T>>` is always weird if you ask me...

I think the way we have it now is fine, I wouldn't worry too much about maintaining compatibility on this particular point

File: independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/core/parameters/converters/OptionalConverter.java [hunk 1]
```diff
@@ -18,7 +19,14 @@ public Object convert(Object parameter) {
         if (parameter == null) {
             return Optional.empty();
         } else if (delegate != null) {
-            return Optional.ofNullable(delegate.convert(parameter));
+            Object converted = delegate.convert(parameter);
+            if (converted != null
+                    && converted instanceof Collection
+                    && ((Collection) converted).isEmpty()) {
```
Comment (Plawn) [inline independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/core/parameters/converters/OptionalConverter.java:25]: Is it really necessary to check if the collection is Empty ?

Comment (Sgitario) [reply independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/core/parameters/converters/OptionalConverter.java:25]: To return optional empty when the collection is empty, yes. See my comment here: https://github.com/quarkusio/quarkus/pull/24055#issuecomment-1056810692

Comment (Plawn) [reply independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/core/parameters/converters/OptionalConverter.java:25]: Oh I see, you are right, It's better this way

**Comment by michalszynkiewicz** [comment]
Should we support `Optional<List>`, etc in the client too? Hmmm... maybe we already do

**Comment by Sgitario** [comment]
> Should we support `Optional<List>`, etc in the client too? Hmmm... maybe we already do

As far I could see, the rest client reactive extension already supports it.

**Pull request merged**